WIP: Docs draft for integration with DVC (#323)
iterative/mlem.ai: content/docs/gto/get-started-dvc.md

# Get Started DVC

To leverage concepts of Model and Data Registries in a more explicit way, you
can denote the `type` of each output. This will let you browse models and data
separately, address them by `name` in `dvc get`, and eventually, see them in DVC
Studio.

> Let's start with marking an artifact as data or model.

Author: Why? I assumed Data Registry should show `type: data`.

> Let's start with marking an artifact as data or model.

Reviewer: It carries 0 information though, right?
"data of type data", similar to "artifact of type artifact" means the same as not defining type at all. it's the most general thing there is (data even more than artifact maybe?), just super abstract. doubt users will use it that way

> Let's start with marking an artifact as data or model.

Author: Good point. Maybe it's `dataset` instead of `data` then? Or anyways, if after subtracting `plots`, `metrics` and `models` everything that's left (among DVC PL inputs and outputs) is `dataset`, then I guess it's redundant?

> Let's start with marking an artifact as data or model.

Reviewer 2: My 2cs. Data is not abstract for me (it's different vs model in my perception). But in DVC it's not needed. Any `out` w/o a specified type can be considered data. 

I would personally try to simplify all of this - no multiple types initially. only models. We are prematurely generalizing this I think.

> Let's start with marking an artifact as data or model.

Reviewer 3: I think of the terms `Data Registry`, `Model Registry`, and `Artifact Registry`. I like having the type track to those names, so I like `data`.

Would love to keep the idea of data registry support around, haven't thought a lot about using `gto` for that, but certainly have thought about models and binaries.

I would love to use/try `gto` as an `Artifact Registry` for build artifacts - specifically compiled binaries.  It might also be interesting to use as a `Container Registry` - there are lots of solutions in this space like [Jfrog Artifactory](https://jfrog.com/artifactory/), [cloudsmith](https://cloudsmith.com/), [GCP](https://cloud.google.com/build/docs/building/store-artifacts-in-cloud-storage).  If you wanted to be really fancy you could support some of those offerings as backends.

With all of that said I think prioritizing the `Model Registry` use case makes a lot of sense, a `Data Registry` would be my next priority.  The solution seems like it may be general enough to support an `Artifact Registry` and `Container Registry`, might be worth doing some thinking about it and if it does make sense and there are advantages to keeping that in `gto` then keep that use case in mind while making the `Model Registry` experience awesome.

> Let's start with marking an artifact as data or model.

Reviewer 3: Trying to think of advantages of having `gto` also be a `Container Registry` and remembered that `mlem`  can deploy docker containers, it would be nice to have the versioning of those container artifacts in `gto` and be able to reference with `gto` syntax.

> Let's start with marking an artifact as data or model.

Reviewer 4: If I might chip in as a user - I absolutely would like GTO to be used also to build a data(set) registry (together with DVC) and possibly a combined data(set) and model registry.

In fact, we intend to use GTO and DVC to build a dataset registry for one (fairly large) client in the coming weeks. By the way, while stuff like MLFlow is a viable alternative to the GTO+MLEM-based model registry (it does not have all of its features but has some others), I don't know any open source alternative to a GTO+DVC-based dataset registry (and general I've yet to see data versioning done better than with DVC)...which means it is a very good selling point IMO.

I like the idea of viewing GTO as a tool to build pretty much any artifact registry, though models and datasets are the most likely use-cases.

If you're using `dvc add` to track your artifact, you'll need to run:

```dvc
$ dvc add mymodel.pkl --type model
```

If you're producing your models in DVC pipeline, you'll need to add
`type: model` to `dvc.yaml` instead:

> 

Author: Now I think there is a user case this change may not support that well. One of our prospects asked to allow a single file (let's say `mymodel.pkl`) to be referenced as several GTO models (e.g. `model1` and `model2` - these are `name`s). Since moving to DVC makes `path` essential (instead of `name`), I don't see how that feature would fit here. 🤔 

The motivation is to be able to promote `model1` and `model2` to different stages at different moments of time separately. To clarify, let's assume there are two populations `mymodel.pkl` should be applied for. You can create stages like `populationA-prod`, `populationA-staging` and `populationB-prod`, `populationB-staging`, if you have many populations, this would make things cumbersome. The solution was to introduce `model1` (for `populationA`) and `model2` (for `populationB`). That required this feature.

The only workaround I see now is to create a "mirror file" with `cp mymodel.pkl mymodel-for-populationB.pkl` in some DVC PL stage. Or keep this `name`:`path` mapping outside of DVC somehow. Do you see any other solutions guys? WDYT?

Reviewer: [related](https://github.com/iterative/mlem.ai/pull/323#discussion_r1131390110)

Reviewer 5: Agree with @omesser that it's related the discussion above. Take a look at the [top-level plots schema](https://dvc.org/doc/user-guide/project-structure/dvcyaml-files#plots), where plots may be identified by either path or an arbitrary name. Feels like following a similar syntax may be best here.

```yaml
stages:
  train:
    cmd: python train.py
    deps:
      - data.xml
    outs:
      - mymodel.pkl:

>           type: model # like this

Reviewer 2: to make it symmetrical with plots, params, etc- why don't we make it:

```yaml
models:
   mymodel.pkl
```
do we plan / need other artifact types? (data can be covered by `outs` I think)

Author: QQ: in this case, do you assume other fields can be specified like:
```yaml
models:
   mymodel.pkl:
     name: particles-classifier
     description: best hits
     labels:
        - boson
        - higgs
```
?

For now it looks like `type: model` may be enough. I can introduce this `models: ` section OFC (and I'd like us to do this, since `artifacts.yaml` is almost like it, which will make a transition easier for existing users with something like `models: artifacts.yaml` - we also have a user asking if we could allow this to have a separate file), but then we need to take into account how we deal with duplication/overwrites if we allow `models` to be specified in `models: ` section and in outs (`type: model`) simultaneously.

Reviewer 5: A top-level `dvc.yaml` section means we need to do some more work on the DVC side instead of using what we have.

If we go this direction, the biggest question is whether to call it `models:` or make it generic like `registry:` or `artifacts:`. The appeal of `models` is that we start small, but I worry it doesn't make much sense long-term because:
1. There's no model-specific functionality in DVC now or in current plans.
2. People already use DVC as a data registry, and it's inevitable people will want to log non-model artifacts.
3. It provides flexibility for custom types (not sure if this is good or bad).

Author: Thanks for ideas. I've updated the doc - please TAL.

Reviewer 5: I think the `dvc.yaml` schema can look like this:

```
# dvc.yaml
stages:
  train:
    cmd: python train.py
    deps:
      - data.xml
    outs:
      - models/mymodel.pkl
artifacts:
  - def-detector:
      path: models/mymodel.pkl
      type: model
      desc: glass defect image classifier
      labels:
        - algo=cnn
        - owner=aguschin
        - project=prod-qual-002
  - models/othermodel.pkl # If no path provided, use name as the path
```
 
This would reuse most of what we already have between:
1. Existing metadata fields for type, desc, labels.
2. Top-level fields we have for plots, metrics, and params.
3. GTO artifacts.yaml (artifacts are in a dict format in artifacts.yaml but a list format above; we could support both to make it less error-prone, which we already do for top-level plots).

I think we can mostly leave alone everything already implemented and only add this new top-level section. I would suggest that only things listed in `artifacts:` (and in artifacts.yaml if we keep it) show up in the Studio registry. Until we get to a non-model registry, we could hide artifacts without `type: model`.

Reviewer 2: It's in the right direction,I like it. by I would make it even simpler initially and do the first class citizen top level `models` section. I don't see a reason to overgeneralize everything at the moment tbh.

Reviewer 5: I think there are good reasons to generalize types. Besides those mentioned above in this thread and by users, we already support `type` in DVC metadata and GTO `artifacts.yaml`, so it fits with what we have already. Is there a good enough reason to rebuild to make it model-specific?

```

You can also specify that while using DVCLive:

```py
live.log_artifact(artifact, "path", type="model")
```

> This will make them appear in DVC Model Registry:

Reviewer: ```suggestion
This will make them appear in [Studio Model Registry](https://dvc.org/doc/studio/user-guide/model-registry/what-is-a-model-registry):
```

Author: Thanks! Overall, let's keep the review scope to the level of ideas and user experience. We don't even know if this will be a separate page in DVC docs, or maybe we integrate it with some other page.

Reviewer: yeah this is a nitpick 😉 but was hard for me to pass the opportunity and suggest

![](https://user-images.githubusercontent.com/6797716/223443152-84f57b79-3395-4965-97f9-edc81896a1dc.png)

and make them shown as models in `dvc ls`:

```dvc
$ dvc ls --registry  # add `--type model` to see models only
 Path           Name                   Type     Labels                       Description
 mymodel.pkl                           model
 data.xml       stackoverflow-dataset  data     data-registry,get-started    imported code
 data/data.xml  another-dataset        data     data-registry,get-started    imported

> ```

Reviewer 5: Does Studio need this, or is solely to provide a CLI option to view the registry? I don't think the latter needs to be high priority unless I'm missing some use case where you need to access it from the CLI.

Not sure whether it can fit into `dvc ls` since the output is quite different (and potentially so are the arguments like `--type`). Need to think about whether we need this and where it can fit.

Reviewer 5: By the way, do we define an artifact as any output that has a `type`?

Author: > some use case where you need to access it from the CLI 

If user would like to download a model locally, but not quite sure which one at the moment, he might want to see this. E.g. I don't remember the model name, but know labels or remember description. This OFC can be solved via Studio, and if we want to push users for that, that's also a decision.

Another use case would be if you're investigating a repo that's not familiar to you (let's say your team has few repos or you look at another team's repo). Again, if we want to make people go to Studio every time for this, it a valid workflow, but IMO it makes you leave CLI and do extra things which can't be inconvenient.

> By the way, do we define an artifact as any output that has a `type`?

Either that, or any input/output file DVC keeps track of can be an artifact (without `type` in it's not defined). I think the latter is simpler and easier to convey. We can call it "file" instead of "artifact" I guess (if we're not going to introduce "compound" artifacts as we discussed before which I don't think is the case. Let's probably don't discuss this though, it's unrelated to this PR and not necessary at all now I believe).

Reviewer 5: Sounds good. Not sure I see enough to make it a p1 yet. WDYT?

Reviewer 5: > > By the way, do we define an artifact as any output that has a `type`?
> 
> Either that, or any input/output file DVC keeps track of can be an artifact (without `type` in it's not defined). I think the latter is simpler and easier to convey.

It might also depend on the schema discussion below. If we have a `model/registry/artifacts` section of `dvc.yaml`, I guess we will only include what's specified there.

The same way you specify `type`, you can specify `description`, `labels` and
`name`. Defining human-readable `name` (should be unique) is useful when you
have complex folder structures or if you artifact can have different paths

> during the project lifecycle.

Reviewer: I think here we need 1 (concise) example setting all relevant fields
e.g.
```bash
dvc add models/mymodel.pkl --name def-detector --type model --description "glass defect image classifier" --label "algo=cnn" --label "owner=aguschin" --label "project=prod-qual-002"
```

You can use `name` to address the object in `dvc get`:

```dvc
$ dvc get $REPO stackoverflow-dataset -o data.xml
```

Now, you usually need a specific model version rather than one from the `main`
branch. You can keep track of the model's lineage by
[registering Semantic versions and promoting your models](/doc/gto/get-started)
(or other artifacts) to stages such as `dev` or `production` with GTO. GTO

> operates by creating Git tags such as `mymodel@v1.2.3` or `mymodel#prod`.

Reviewer 5: I thought the stage tags always had some number like `mymodel#prod#1`?

Author: It can be `mymodel#prod` as well. It's called "simple" Git tag format and it's not the default one. https://mlem.ai/doc/gto/user-guide/#git-tags-format

Knowing the right Git tag, you can get the model locally:

```dvc
$ dvc get $REPO mymodel.pkl --rev mymodel@v1.2.3
```

Check out
[GTO User Guide](/doc/gto/user-guide/#getting-artifacts-in-systems-downstream)
to learn how to get the Git tag of the `latest` version or version currently
promoted to stages like `prod`.

## Getting models in CI/CD

Git tags are great to [kick off CI/CD](/doc/gto/user-guide/#acting-in-cicd)
pipeline in which we can consume our model. You can use
[GTO GitHub action](https://github.com/iterative/gto-action) to interpret the
Git tag that triggered the workflow and act based on that. If you simply need to
download the model to CI, you can also use this Action with `download` option:

```yaml
steps:
  - uses: actions/checkout@v3
  - id: gto
    uses: iterative/gto-action@v1
    with:

>       download: True # you can provide a specific destination path here instead of `True`

Reviewer 5: Can you give an example? Will it download all artifacts in the repo?

Author: This should download the artifact, e.g. it will run `dvc get . mymodel --rev $GITHUB_REF` for a Git tag `mymodel@v0.0.1`.

```

## Restricting which types are allowed

To specify which `type`s are allowed to be used, you can add the following to
your `.dvc/config`:

```
# .dvc/config
types: [model, data]

> ```

Reviewer 5: Do you think it needs to be part of the initial scope? I know it was requested, but it doesn't feel strictly necessary yet.

Author: Not necessary. I'll make a note in the docs page.

> ## Seeing new model versions pushed with DVC experiments

Reviewer: I didn't, I'm a bit concerned for implicit behaviors, we should probably find a way to give the user control over what to do with which artifacts on exp push. wdyt?

> ## Seeing new model versions pushed with DVC experiments

Author: What example of implicit behavior you have in mind? Like `pushing` a model that can be few GB in size? Not quite have specific examples in mind.

> ## Seeing new model versions pushed with DVC experiments

Reviewer: yes, auto-pushing, exactly. maybe auto-versioning as well in the future (could be useful if running in the pipeline CI-CD as part of release. generate model, push it, and assign a version using GTO

After you run `dvc exp push` to push your experiment that updates your model,

> you'll see a commit candidate to be registered:

Reviewer 5: Do you think we should allow registration of unmerged experiments? Or maybe restrict what actions are available for them?

Author: Good question. Don't have a strong opinion.
First, it's possible to do, so why not. We can also allow to click on "register", but then say something like "We advise to merge the experiment first" with buttons like "create a PR in GH" (default) and "register anyway".
We can prohibit registering (again, don't see a reason except for skipping polluting repo with dangling refs in a the workflow that requires users to merge experiments first).
We can delay answering this question for now I think.

# Get Started DVC

To leverage concepts of Model and Data Registries in a more explicit way, you
can denote the `type` of each output. This will let you browse models and data
separately, address them by `name` in `dvc get`, and eventually, see them in DVC
Studio.

Let's start with marking an artifact as data or model.

If you're using `dvc add` to track your artifact, you'll need to run:

```dvc
# note that all CLI options are optional:
$ dvc add models/mymodel.pkl \
    --type model \  # this makes DVC understand it's an ML model
    --name def-detector \
    --description "glass defect image classifier" \
    --label "algo=cnn" \
    --label "owner=aguschin" \
    --label "project=prod-qual-002"
```

<details>

### Beside tracking this as usually, his will add it to a top section called `registry` in your `dvc.yaml`

```yaml
# dvc.yaml
registry:
  def-detector: # just like with plots, this could be a path or any string ID
    type: model
    description: glass defect image classifier
    labels:
      - algo=cnn
      - owner=aguschin
      - project=prod-qual-002

>     path: models/mymodel.pkl # specify path if use alias to name this

Author: ```suggestion
    path: models/mymodel.pkl # specify path if using alias to reference the artifact
```

```

If you want this to be in a separate file (say, `artifacts.yaml`), you can tell
DVC to use it with:

```yaml
# dvc.yaml
registry: artifacts.yaml

> ```

Author: Extra for now, but was requested by some users

Reviewer 3: Love this!  I would love to do this with `plots` too.
